File: schemas/ui/registrationSchema.ts
import { FieldSchema } from "../fieldSchema.interface";

export const uiRegistrationSchema: FieldSchema = {
  schemaId:    "ui-registration",
  version:     "1.0.0",
  target:      "ui",
  description: "Customer registration form — full hierarchical schema example",

  goalCondition: {
    ui: {
      successSelector: ".success-message, .alert-success, [data-success]",
      successText:     "successfully",
      urlContains:     "/success",
      waitForSelector: "form",
    },
  },

  astarConfig: {
    maxIterations:    500,
    heuristicWeight:  1.0,
    goalTimeout:      30000,
    allowPartialGoal: false,
  },

  sections: [

    // ═══════════════════════════════════════════════════════
    // SECTION 1: Personal Info — Mandatory
    // ═══════════════════════════════════════════════════════
    {
      sectionName: "personalInfo",
      mandatory:   true,
      weight:      10,
      label:       "Personal Information",
      stepIndex:   0,
      fields: [
        {
          name:        "firstName",
          type:        "text",
          weight:      9,
          label:       "First Name",
          placeholder: "Enter your first name",
          selector:    '[name="firstName"], [id="firstName"], [placeholder*="first" i]',
          validationRule: {
            pattern:      "^[a-zA-Z]{2,50}$",
            minLength:    2,
            maxLength:    50,
            errorMessage: "First name: 2-50 alphabetic characters only",
          },
          dataHints: {
            validSamples:   ["Karthick", "Priya", "John", "Aisha"],
            invalidSamples: ["A", "123Name", ""],
          },
        },
        {
          name:        "lastName",
          type:        "text",
          weight:      9,
          label:       "Last Name",
          placeholder: "Enter your last name",
          selector:    '[name="lastName"], [id="lastName"], [placeholder*="last" i]',
          validationRule: {
            pattern:      "^[a-zA-Z]{2,50}$",
            minLength:    2,
            maxLength:    50,
            errorMessage: "Last name: 2-50 alphabetic characters only",
          },
          dataHints: {
            validSamples:   ["Kumar", "Smith", "Raj", "Chen"],
            invalidSamples: ["", "1", "A"],
          },
        },
        {
          name:        "email",
          type:        "email",
          weight:      10,
          label:       "Email Address",
          placeholder: "Enter your email address",
          selector:    '[name="email"], [type="email"], [id="email"]',
          validationRule: {
            pattern:      "^[\\w.-]+@[\\w.-]+\\.\\w{2,}$",
            errorMessage: "Please enter a valid email address",
          },
          dataHints: {
            validSamples:   ["test@astra.com", "user@example.com", "qa@freewheel.tv"],
            invalidSamples: ["notanemail", "missing@", "@nodomain"],
            format:         "email",
          },
        },
        {
          name:        "phone",
          type:        "phone",
          weight:      8,
          label:       "Mobile Number",
          placeholder: "Enter 10-digit mobile number",
          selector:    '[name="phone"], [type="tel"], [id="phone"]',
          validationRule: {
            pattern:      "^[6-9]\\d{9}$",
            errorMessage: "Enter a valid 10-digit Indian mobile number",
          },
          dataHints: {
            validSamples:   ["9876543210", "8123456789", "7001234567"],
            invalidSamples: ["12345", "0123456789", "abcdefghij"],
            locale:         "en-IN",
          },
        },
      ],
    },

    // ═══════════════════════════════════════════════════════
    // SECTION 2: Address — Mandatory
    // Nested children inherit mandatory: true
    // ═══════════════════════════════════════════════════════
    {
      sectionName: "address",
      mandatory:   true,
      weight:      8,
      label:       "Address Details",
      stepIndex:   0,
      fields: [
        {
          name:     "addressBlock",
          type:     "text",
          weight:   8,
          label:    "Address",
          selector: "[data-section='address'], fieldset.address",
          children: [
            {
              name:        "street",
              type:        "text",
              weight:      8,
              label:       "Street / Area",
              placeholder: "Enter street name or area",
              selector:    '[name="street"], [name="address1"], [id="street"]',
              validationRule: {
                minLength:    3,
                maxLength:    100,
                errorMessage: "Street address is required",
              },
              dataHints: {
                validSamples:   ["12 Main Street", "45 Gandhi Road", "Plot 7 Lake View"],
                invalidSamples: ["", "A"],
              },
            },
            {
              name:        "city",
              type:        "text",
              weight:      8,
              label:       "City",
              placeholder: "Enter city name",
              selector:    '[name="city"], [id="city"]',
              validationRule: {
                pattern:      "^[a-zA-Z\\s]{2,50}$",
                errorMessage: "Enter a valid city name",
              },
              dataHints: {
                validSamples:   ["Chennai", "Bangalore", "Mumbai", "Hyderabad"],
                invalidSamples: ["", "12345"],
              },
            },
            {
              name:        "state",
              type:        "dropdown",
              weight:      7,
              label:       "State",
              selector:    'select[name="state"], [id="state"]',
              validValues: [
                "Tamil Nadu", "Karnataka", "Maharashtra",
                "Delhi", "Telangana", "Kerala",
              ],
              dataHints: {
                validSamples: ["Tamil Nadu", "Karnataka"],
              },
            },
            {
              name:        "pincode",
              type:        "number",
              weight:      7,
              label:       "Pincode",
              placeholder: "Enter 6-digit pincode",
              selector:    '[name="pincode"], [name="zipcode"], [id="pincode"]',
              validationRule: {
                pattern:      "^[1-9][0-9]{5}$",
                errorMessage: "Enter a valid 6-digit pincode",
              },
              dataHints: {
                validSamples:   ["600001", "560001", "400001"],
                invalidSamples: ["12345", "000000", "abcdef"],
              },
            },
          ],
        },
      ],
    },

    // ═══════════════════════════════════════════════════════
    // SECTION 3: Credentials — Mandatory | Step 2
    // ═══════════════════════════════════════════════════════
    {
      sectionName: "credentials",
      mandatory:   true,
      weight:      10,
      label:       "Login Credentials",
      stepIndex:   1,
      fields: [
        {
          name:        "username",
          type:        "text",
          weight:      10,
          label:       "Username",
          placeholder: "Choose a username",
          selector:    '[name="username"], [id="username"]',
          validationRule: {
            pattern:      "^[a-zA-Z0-9_]{4,20}$",
            minLength:    4,
            maxLength:    20,
            errorMessage: "Username: 4-20 alphanumeric characters or underscore",
          },
          dataHints: {
            validSamples:   ["astra_user", "test_qa_01", "karthick_001"],
            invalidSamples: ["ab", "user name", "a".repeat(21)],
            unique:         true,
          },
        },
        {
          name:        "password",
          type:        "password",
          weight:      10,
          label:       "Password",
          placeholder: "Create a strong password",
          selector:    '[name="password"], [id="password"]',
          validationRule: {
            pattern:      "^(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$",
            minLength:    8,
            errorMessage: "Min 8 chars with 1 uppercase, 1 number, 1 special char",
          },
          dataHints: {
            validSamples:   ["Test@1234", "Secure#Pass9", "Astra@2024!"],
            invalidSamples: ["weak", "12345678", "NoSpecial1"],
          },
        },
        {
          name:        "confirmPassword",
          type:        "password",
          weight:      10,
          label:       "Confirm Password",
          placeholder: "Re-enter your password",
          selector:    '[name="confirmPassword"], [name="confirm_password"]',
          dependencies: [
            {
              dependsOn: "password",
              whenValue: "",
              operator:  "notEquals",
            },
          ],
          validationRule: {
            errorMessage: "Passwords must match",
          },
          dataHints: {
            validSamples: ["Test@1234"],
          },
        },
      ],
    },

    // ═══════════════════════════════════════════════════════
    // SECTION 4: Preferences — Optional
    // A* skips this entire section
    // ═══════════════════════════════════════════════════════
    {
      sectionName: "preferences",
      mandatory:   false,
      weight:      3,
      label:       "Preferences (Optional)",
      stepIndex:   1,
      fields: [
        {
          name:        "gender",
          type:        "dropdown",
          weight:      3,
          label:       "Gender",
          selector:    'select[name="gender"], [id="gender"]',
          validValues: ["Male", "Female", "Non-binary", "Prefer not to say"],
          dataHints: {
            validSamples: ["Male", "Female"],
          },
        },
        {
          name:        "dateOfBirth",
          type:        "date",
          weight:      3,
          label:       "Date of Birth",
          placeholder: "YYYY-MM-DD",
          selector:    '[name="dateOfBirth"], [name="dob"], [id="dateOfBirth"]',
          dataHints: {
            validSamples: ["1995-06-15", "2000-01-01"],
            format:       "YYYY-MM-DD",
          },
        },
        {
          name:     "newsletterOptIn",
          type:     "checkbox",
          weight:   2,
          label:    "Subscribe to Newsletter",
          selector: '[name="newsletter"], [id="newsletterOptIn"]',
          validValues: ["true", "false"],
          dataHints: {
            validSamples:   ["true"],
            invalidSamples: ["false"],
          },
        },
      ],
    },

  ],
};

File: schemas/ui/registration_schema.py
"""ASTRA manual UI registration schema."""
from __future__ import annotations

from schemas.field_schema import (
    AStarConfig,
    DataHints,
    Field,
    FieldSchema,
    GoalCondition,
    Section,
    ValidationRule,
)
from utils.env_loader import ENV


def get_registration_ui_schema() -> FieldSchema:
    return FieldSchema(
        id="ui-registration-v1",
        name="User Registration UI",
        version="1.0.0",
        type="ui",
        base_url=ENV.BASE_URL,
        target_endpoint=ENV.TARGET_PAGE_URL,
        sections=[
            Section(
                id="personalInfo",
                name="Personal Information",
                order=1,
                fields=[
                    Field(
                        name="firstName",
                        type="text",
                        label="First Name",
                        selector="#firstName, [name=firstName], [placeholder*='First']",
                        required=True,
                        priority=9,
                        validation_rules=[
                            ValidationRule(type="minLength", value=2),
                            ValidationRule(type="maxLength", value=50),
                        ],
                        data_hints=DataHints(
                            valid_examples=["John", "Jane"],
                            invalid_examples=["J", "123"],
                        ),
                    ),
                    Field(
                        name="lastName",
                        type="text",
                        label="Last Name",
                        selector="#lastName, [name=lastName], [placeholder*='Last']",
                        required=True,
                        priority=9,
                        validation_rules=[
                            ValidationRule(type="minLength", value=2),
                            ValidationRule(type="maxLength", value=50),
                        ],
                        data_hints=DataHints(
                            valid_examples=["Doe", "Smith"],
                            invalid_examples=["D", "456"],
                        ),
                    ),
                    Field(
                        name="email",
                        type="email",
                        label="Email",
                        selector="#email, [name=email], [type=email]",
                        required=True,
                        priority=10,
                        validation_rules=[
                            ValidationRule(type="pattern", value="^[^@]+@[^@]+\\.[^@]+$"),
                        ],
                        data_hints=DataHints(
                            valid_examples=["user@example.com"],
                            invalid_examples=["notanemail"],
                            format="email",
                        ),
                    ),
                    Field(
                        name="phone",
                        type="tel",
                        label="Phone",
                        selector="#phone, [name=phone], [type=tel]",
                        required=False,
                        priority=5,
                        data_hints=DataHints(
                            valid_examples=["+1234567890"],
                        ),
                    ),
                ],
            ),
            Section(
                id="credentials",
                name="Account Credentials",
                order=2,
                fields=[
                    Field(
                        name="username",
                        type="text",
                        label="Username",
                        selector="#username, [name=username], [placeholder*='Username']",
                        required=True,
                        priority=10,
                        validation_rules=[
                            ValidationRule(type="minLength", value=3),
                            ValidationRule(type="maxLength", value=30),
                        ],
                        data_hints=DataHints(
                            valid_examples=["john_doe", "alice123"],
                        ),
                    ),
                    Field(
                        name="password",
                        type="password",
                        label="Password",
                        selector="#password, [name=password], [type=password]",
                        required=True,
                        priority=10,
                        validation_rules=[
                            ValidationRule(type="minLength", value=8),
                        ],
                        data_hints=DataHints(
                            valid_examples=["Secure@123"],
                            invalid_examples=["short"],
                        ),
                    ),
                    Field(
                        name="confirmPassword",
                        type="password",
                        label="Confirm Password",
                        selector="#confirmPassword, [name=confirmPassword], [placeholder*='Confirm']",
                        required=True,
                        priority=9,
                        data_hints=DataHints(
                            valid_examples=["Secure@123"],
                        ),
                    ),
                ],
            ),
            Section(
                id="preferences",
                name="Preferences",
                order=3,
                fields=[
                    Field(
                        name="country",
                        type="select",
                        label="Country",
                        selector="#country, [name=country], select[id*='country']",
                        required=True,
                        priority=8,
                        data_hints=DataHints(
                            enum_values=["US", "GB", "CA", "AU", "IN"],
                        ),
                    ),
                    Field(
                        name="termsAccepted",
                        type="checkbox",
                        label="Accept Terms",
                        selector="#terms, [name=terms], [name=termsAccepted]",
                        required=True,
                        priority=6,
                        data_hints=DataHints(
                            valid_examples=["true"],
                            invalid_examples=["false"],
                        ),
                    ),
                ],
            ),
        ],
        goal_conditions=[
            GoalCondition(type="url_contains", value="/success"),
            GoalCondition(type="element_visible", value=".success-message, #success, [data-testid=success]"),
        ],
        astar_config=AStarConfig(
            max_iterations=500,
            heuristic_weight=1.0,
            allow_optional=False,
            timeout_ms=30000,
        ),
    )


REGISTRATION_UI_SCHEMA = get_registration_ui_schema()
